File: 7.reverse-integer.py
#
# @lc app=leetcode id=7 lang=python3
#
# [7] Reverse Integer
#

# @lc code=start
class Solution:
    def reverse(self, x: int) -> int:
        MAX_INT = 2**31 - 1
        MIN_INT = -2**31
        
        flag = -1 if x < 0 else 1
        x = abs(x)

        result = 0
        while x != 0:
            remain = x % 10
            x //= 10

            result = result * 10 + remain
        
        result *= flag
        if result < MIN_INT or result > MAX_INT:
            return 0
        return result

# @lc code=end

"""
`s = (x > 0) - (x < 0)`
센스 ㄷㄷ
"""

print(Solution.reverse(None, 1563847412))

File: Top Interview Questions/src/7.reverse-integer.js
/*
 * @lc app=leetcode id=7 lang=javascript
 *
 * [7] Reverse Integer
 */

// @lc code=start
/**
 * @param {number} x
 * @return {number}
 */
var reverse = function(x) {
  const MIN = -(2 ** 31);
  const MAX = 2 ** 31 - 1;

  var flag = 1;
  if (x < 0) {
    var flag = -1;
    x *= -1;
  }

  var answer = 0;
  while (x > 0) {
    var num = x % 10;
    x = parseInt(x / 10);

    answer = (answer * 10) + num;
    if (answer < MIN || answer > MAX) {
      return 0;
    }
  }

  return answer * flag;
};
// @lc code=end

